### PR TITLE
Create a WatchStatus class for the high-level REST client.

### PR DESCRIPTION
**Pull request by jtibshirani**
This class will be used in a few of the watcher responses ('get watch', 'ack watch', etc.), so it's being introduced first in its own PR.

To create this PR, I first copied over all `WatchStatus` and its dependencies without modification, then iteratively cleaned them up. It's likely easiest to review by going through the commits separately.

### CODE REVIEW

File: client/rest-high-level/src/main/java/org/elasticsearch/client/watcher/WatchStatusDateParser.java
Comment (jtibshirani) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/watcher/WatchStatusDateParser.java]: Somewhat confusingly, the timestamps in `WatchStatus` are subject to different rules than the ones in `ActionStatus`: in `WatchStatus` they can be longs, and are always converted to UTC, but in `ActionStatus` they must be ISO dates and retain their original timezone. For consistency, I've aimed to keep this logic while refactoring.

File: client/rest-high-level/src/test/java/org/elasticsearch/client/watcher/WatchStatusTests.java
Comment (jtibshirani) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/watcher/WatchStatusTests.java]: In order to sanity check this PR, I manually added some basic parsing tests. One other option would be to keep the `toXContent` logic, and create a `AbstractXContentTestCase`. I decided against that approach because there was quite a bit of logic associated with `toXContent`, and this isn't the test direction we're ultimately aiming for.

Comment (romseygeek) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/watcher/WatchStatusTests.java]: I think you can avoid that by overriding `AbstractXContentTestCase#assertToXContentEquivalence`?  I think it's worth using `AbstractXContentTestCase` here, it's going to be much more thorough than hand-rolling parsing tests.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/watcher/ActionStatus.java
Comment (romseygeek) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/watcher/ActionStatus.java]: I think using `ConstructingObjectParser` would make all these parse methods much clearer

**Review by romseygeek — COMMENTED** [review]
This looks a great start @jtibshirani, I left a couple of comments.

File: client/rest-high-level/src/test/java/org/elasticsearch/client/watcher/WatchStatusTests.java
Comment (hub-cap) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/watcher/WatchStatusTests.java]: we need to come to consensus on that, since we wont actually have the toXContent on a lot of classes, and i dont know if i want to add it solely for completing a test. Also, testing that the class can ser/deser the bytes is one thing, but testing that a server side response ser's into something that the client can deser is much more valuable, since thats the way we know the client actually works. After support summit ill have a few brain cells to devote to this as Id like to get an answer out sooner than later.

**Review by iverase — COMMENTED** [review]
Thanks Julie for this super effort. I am wondering if this classes should implement {toXContent} for Json serialisation and potentially {Writable} as well.

File: client/rest-high-level/src/test/java/org/elasticsearch/client/watcher/WatchStatusTests.java
Comment (jtibshirani) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/watcher/WatchStatusTests.java]: Thanks everyone for taking a look. I purposefully removed the `toXContent` method and stream serialization methods (in the commit “Remove logic related to writing xContent and stream serialization”) because in the high-level REST client, response classes are only ever deserialized from xContent. As @hub-cap mentioned, we expect this to be the case for most of the response classes.

In tests like `AbstractXContentTestCase`, we test parsing by creating an object, generating xContent, then parsing it back to an object and checking equality. So because we no longer have a `toXContent` method, this testing strategy doesn’t apply. My understanding was that we did not want to keep around `toXContent` just to help with testing, and would prefer to move in a direction where we’re testing against a live server (or some kind of API spec?)

@hub-cap do you have thoughts about the right testing approach for this PR? It would be nice make progress on getting this merged (even if the testing approach is temporary), so others can continue their watcher HLRC work.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/watcher/ActionStatus.java
Comment (jtibshirani) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/watcher/ActionStatus.java]: Thanks @romseygeek -- I agree, but I opted not to make major refactors to the shared parsing logic to keep this PR focused. Maybe I could do a follow-up to clean up `ActionStatus` (both the HLRC version, and the core watcher version)?

File: client/rest-high-level/src/test/java/org/elasticsearch/client/watcher/WatchStatusTests.java
Comment (hub-cap) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/watcher/WatchStatusTests.java]: I need to sit down to make a approach, ill finally have time early next week. Id prefer the manual parsing in order to test the fromXContents, for now.

Comment (jtibshirani) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/watcher/WatchStatusTests.java]: I clarified with @hub-cap and he thinks we should stick with this testing approach for this PR + proceed with the code review (or as he says, :shipit: :) )

Comment (iverase) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/watcher/WatchStatusTests.java]: There are already a couple of actions implemented. For example this one which is the delete watcher action:

https://github.com/elastic/elasticsearch/pull/32337

Responses extends the abstract class ActionResponse which seems to require serialisation/deserialisation. The responses classes implement ToXContentObject as well  so no sure if for consistency we should do the same.

Comment (jtibshirani) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/watcher/WatchStatusTests.java]: Right, but looking at a more recent xpack HLRC PR (#32332), I saw that the responses no longer contained the extraneous serialization/ deserialization logic (see `PutUserResponse`). I assumed that this was the new preferred way, especially now that we are moving the responses into their own lightweight module. Tagging @hub-cap in case my assumptions are off!

Comment (hub-cap) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/watcher/WatchStatusTests.java]: @jtibshirani is correct

Comment (iverase) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/watcher/WatchStatusTests.java]: This other approach looks fine to me as well. Thanks for the clarification!

**Comment by romseygeek** [comment]
I think this is ready to merge now?  If there's consensus on the tests, then I'm +1

**Comment by jtibshirani** [comment]
👍 I'll rebase and make sure the tests are green. It would be great to get a PR approval, if everyone is on board.

**Review by iverase — APPROVED** [review]
So here is my thumb up. 

We agreed that implementation of request/response will follow the implementation of  #32332 and therefore no need of serialisation/deserialisation of WatchStatus. 
LGTM

**Comment by hub-cap** [comment]
:shipit:

**Pull request merged**
